docs/graph_engine/dataset.rst

```diff
@@ -0,0 +1,201 @@
+
+We convert this dataset to a data pipeline by splitting it into windows. Each window is effectively a separate
+dataset object and the dataset pipeline consists of multiple windows which each include mulitple blocks / batches.
+When we run a function on a dataset pipeline it is not executed immediately, instead it is staged and only run
+per window when required when iterating over the dataset.
+More about dataset pipelines, `here <https://docs.ray.io/en/latest/data/pipelining-compute.html#pipelining-datasets>`_.
+
+* Higher parallelism is generally better for performance. Lower blocks per window means lower latency but gives less room for concurrent tasks.
+* Cluster memory should be 2-5x the window size to avoid spilling, you can see window size by using `dataset.stats()`.
+
+.. code-block:: python
+
+    >>> pipe = dataset.window(blocks_per_window=2)
+    >>> pipe
+    DatasetPipeline(num_windows=3, num_stages=1)
+
+In order to rerun this dataset multiple times, one per epoch, we use the repeat command.
+In this example we call repeat before running any transforms on the dataset, therefore the transform outputs will not be cached between epochs.
+If repeat is run after a transform, the result of the transform will be cached, `more here <https://docs.ray.io/en/latest/data/advanced-pipelines.html#dataset-pipeline-per-epoch-shuffle>`_.
+
+.. code-block:: python
+
+    >>> n_epochs = 2
+    >>> pipe = pipe.repeat(n_epochs)
+    >>> pipe
+    DatasetPipeline(num_windows=6, num_stages=1)
+
+We add shuffling at this part of the dataset so that we only shuffle node ids, not the whole query return. It is important to add this after repeat so it does not get cached.
+
+    >>> pipe = pipe.random_shuffle_each_window(seed=0)
+    >>> pipe
+    DatasetPipeline(num_windows=6, num_stages=2)
+
+Use `map_batches <https://docs.ray.io/en/latest/data/api/dataset.html#ray.data.Dataset.map_batches>`_
+to map node ids from the sampler to a dictionary of node features and labels for the model forward function.
+Since this is run on the dataset pipeline, the node ids will not be mapped all at once, only when needed during iteration.
+
+For each query output vector, all first dimensions need to be equal to the batch size.
+
+.. code-block:: python
+
+    >>> def transform_batch(idx: list) -> dict:
+    ...     return {"features": g.node_features(idx, np.array([[0, 50]]), feature_type=np.float32), "labels": g.node_features(idx, np.array([[1, 1]]), feature_type=np.float32)}
+    >>> pipe = pipe.map_batches(transform_batch)
+    >>> pipe
+    DatasetPipeline(num_windows=6, num_stages=3)
+
+Finally we iterate over the dataset `n_epochs` times.
+
+.. code-block:: python
+
+    >>> epoch_iter = pipe.iter_epochs()
+    >>> epoch_pipe = next(epoch_iter)
+    >>> batch = next(epoch_pipe.iter_torch_batches(batch_size=2))
+    >>> batch
+    {'features': tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
```

Comment:
nit: just check first few zeros and then wildcard the rest to keep docs concise.